## Phoneme Accuracy

In [1]:
import json 
def load_json(path):
    """
    Load a json file and return the content as a dictionary.
    """
    import json

    with open(path, "r", encoding= 'utf-8') as f:
        data = json.load(f)
    return data

def get_unique_tokens(data):
    """
    Get unique tokens from the dataset.
    """
    unique_tokens = set()
    for item in data:
        transcription = item["text"]
        for token in transcription.split():
            unique_tokens.add(token)
    return list(unique_tokens)

testset_path= "../dataset/LSVSC_test_word.json"
vocab_train_path = "../dataset/word_vocab_lsvsc.json"

vocab_train = load_json(vocab_train_path)
unique_token_test = get_unique_tokens(load_json(testset_path))
oov_word = []
for token in unique_token_test:
    if token not in vocab_train:
        oov_word.append(token)

print(f"Number of OOV words: {len(oov_word)}, Percentage: {len(oov_word)/len(unique_token_test)*100:.2f}%")

Number of OOV words: 116, Percentage: 3.78%


In [33]:


# num <unk> in test result
test_path_result = "./result/result-tasa-w2i-lsvsc.json"
test_result = load_json(test_path_result)
num_unk = 0
idx_list_data = []

for idx, item in enumerate(test_result):
    try:
        transcription = item["gold"]
        for idx_word, token in enumerate(transcription.split()):
            if token == "<unk>":
                num_unk += 1
                idx_list_data.append((idx, idx_word))
    except: 
        continue


# num correct fill by phoneme based model 

test_result_phoneme = load_json("./result/result-tasa-p2i_lsvsc.json")
phoneme_test_path = "../dataset/LSVSC_test_phoneme.json"
phoneme_test_data = load_json(phoneme_test_path)
num_correct_fill = 0

correct_filled = []
cannot_be_filled = []
tried_to_filled_but_wrong = []
test_data = load_json(testset_path)
sample_correct = []
total_num = 0
for idx, idx_word in idx_list_data:
    if idx_word < len(test_result_phoneme[idx]["predicted"].split(' ')):
        gold = test_result_phoneme[idx]["gold"].split(' ')[idx_word]
        pred = test_result_phoneme[idx]["predicted"].split(' ')[idx_word]
        # print(f"Gold: {gold}, Pred: {pred}")
        if gold == pred:
            
            correct_filled.append(gold)
            num_correct_fill += 1
            if test_result_phoneme[idx]['gold'] == test_result_phoneme[idx]["predicted"]:
                print(f"Gold phoneme: {test_result_phoneme[idx]['gold']} | pred phonem: {test_result_phoneme[idx]["predicted"]}")

                sample_correct.append([
                    test_result[idx]['gold'],
                    test_data[idx]['text']
                ])
        else:
            cannot_be_filled.append(gold)
            tried_to_filled_but_wrong.append(pred)
        total_num += 1

print(f"Number of OOV words: {len(oov_word)}, Percentage: {len(oov_word)/len(unique_token_test)*100:.2f}%")
print(f"Number of <unk> in transcript of test dataset: {total_num}")
print(f"Number of correct fill by phoneme based model: {num_correct_fill}")
print(f"Percentage of correct fill by phoneme based model: {num_correct_fill/total_num*100:.2f}%")
print(f"Correctly filled words: {set(correct_filled)}")
print(f"Cannot be filled words: {set(cannot_be_filled)}")
print(f"Tried to fill but wrong predictions: {set(tried_to_filled_but_wrong)}")

Gold phoneme: tu˧˥ kwiən- tʰəː˧˩ ʝaːi˨˩ tʰɯən˨˩ tʰɯən~˧ˀ˩ | pred phonem: tu˧˥ kwiən- tʰəː˧˩ ʝaːi˨˩ tʰɯən˨˩ tʰɯən~˧ˀ˩
Gold phoneme: haːi- taːi- t͡ɕoŋ˧˥ naːŋ̟˧ˀ˩ han˧˥ dɯŋ˧˥ ʝaːŋ˧ˀ˩ haːŋ˧˥ ʈ͡ʂɯəŋ~˧˥ fɔŋ˨˩ laːm˨˩ viəŋ~˧ˀ˩ | pred phonem: haːi- taːi- t͡ɕoŋ˧˥ naːŋ̟˧ˀ˩ han˧˥ dɯŋ˧˥ ʝaːŋ˧ˀ˩ haːŋ˧˥ ʈ͡ʂɯəŋ~˧˥ fɔŋ˨˩ laːm˨˩ viəŋ~˧ˀ˩
Gold phoneme: dem- vɯə˨˩ roi˨˩ mon~˧ˀ˩ ten- ʈ͡ʂɔŋ- daːm˧˥ zɔi˨˩ bɔ˧ˀ˩ daː˧ˀ˥ t͡ɕen~˧˥ ve˨˩ taːi- han˧˥ | pred phonem: dem- vɯə˨˩ roi˨˩ mon~˧ˀ˩ ten- ʈ͡ʂɔŋ- daːm˧˥ zɔi˨˩ bɔ˧ˀ˩ daː˧ˀ˥ t͡ɕen~˧˥ ve˨˩ taːi- han˧˥
Gold phoneme: dem- vɯə˨˩ roi˨˩ mon~˧ˀ˩ ten- ʈ͡ʂɔŋ- daːm˧˥ zɔi˨˩ bɔ˧ˀ˩ daː˧ˀ˥ t͡ɕen~˧˥ ve˨˩ taːi- han˧˥ | pred phonem: dem- vɯə˨˩ roi˨˩ mon~˧ˀ˩ ten- ʈ͡ʂɔŋ- daːm˧˥ zɔi˨˩ bɔ˧ˀ˩ daː˧ˀ˥ t͡ɕen~˧˥ ve˨˩ taːi- han˧˥
Gold phoneme: mɛ˧ˀ˩ vaː˨˩ t͡ɕi˧ˀ˩ nɔ˧˥ daːŋ- ŋoi˨˩ vaː˧˥ lɯəi˧˥ ben- məi˧˥ t͡ɕiəŋ~˧˥ tʰwiən˨˩ tʰuŋ˧˥ voi˧ˀ˩ vaːŋ˨˩ t͡ɕaːi˧ˀ˩ len- t͡ɕo˧ˀ˥ hɯŋ- nam˨˩ | pred phonem: mɛ˧ˀ˩ vaː˨˩ t͡ɕi˧ˀ˩ nɔ˧˥ daːŋ- ŋoi˨˩ vaː˧˥ lɯəi˧˥ ben- məi˧˥ t͡ɕiəŋ~˧˥ tʰwiən˨˩ tʰuŋ˧˥ voi˧ˀ˩ vaːŋ˨˩

In [45]:
duplicate_check = set()
for gold, text in sample_correct:
    if (gold, text) not in duplicate_check:
        print(f"Gold unknown word: {gold} \nPredicted Phoneme Converted: {text} \n{'-'*20}")
        duplicate_check.add((gold, text))




Gold unknown word: tú quyên thở dài <unk> thượt 
Predicted Phoneme Converted: tú quyên thở dài thườn thượt 
--------------------
Gold unknown word: hai tay chống nạnh hắn đứng dạng <unk> trước phòng làm việc 
Predicted Phoneme Converted: hai tay chống nạnh hắn đứng dạng háng trước phòng làm việc 
--------------------
Gold unknown word: đêm vừa rồi một tên trong đám <unk> <unk> đã chết về tay hắn 
Predicted Phoneme Converted: đêm vừa rồi một tên trong đám giòi bọ đã chết về tay hắn 
--------------------
Gold unknown word: mẹ và chị nó đang ngồi vá lưới bên mấy chiếc thuyền <unk> vội vàng chạy lên chỗ hưng nằm 
Predicted Phoneme Converted: mẹ và chị nó đang ngồi vá lưới bên mấy chiếc thuyền thúng vội vàng chạy lên chỗ hưng nằm 
--------------------
Gold unknown word: cũng như những người còn ngờ vực khác trong đám đông anh đã nhìn thấy rõ cái trò <unk> dân của tên chỉ huy trưởng 
Predicted Phoneme Converted: cũng như những người còn ngờ vực khác trong đám đông anh đã nhìn thấy rõ cái trò

In [46]:
def save_json(path, data):
    """
    Save a dictionary to a json file.
    """
    import json

    with open(path, "w", encoding= 'utf-8') as f:
        json.dump(data, f, indent=4, ensure_ascii=False)

save_json("./result/correctly_filled_words.json", list(set(correct_filled)))
save_json("./result/cannot_be_filled_words.json", list(set(cannot_be_filled)))
save_json("./result/tried_to_filled_but_wrong_predictions.json", list(set(tried_to_filled_but_wrong)))

## Pearson correlation

In [53]:

# load unique word vocab 
# load unique word phoneme 

def load_json(path):
    """
    Load a json file and return the content as a dictionary.
    """
    import json

    with open(path, "r", encoding= 'utf-8') as f:
        data = json.load(f)
    return data

def save_json(path, data):
    """
    Save a dictionary to a json file.
    """
    import json

    with open(path, "w", encoding= 'utf-8') as f:
        json.dump(data, f, indent=4, ensure_ascii=False)

from jiwer import process_words

def compute_token_stats_with_jiwer(path, correct_at_least=1, save_path=None):
    data = load_json(path)
    data = [x for x in data if isinstance(x, dict) and "gold" in x and "predicted" in x]

    stats = {}

    for item in data:
        gold = item["gold"].strip()
        pred = item["predicted"].strip()

        out = process_words(gold, pred)

        aligns = out.alignments[0]

        # unwrap references/hypotheses if nested
        ref = out.references[0]     # list[str]

        for ch in aligns:
            op = ch.type
            rs, re = ch.ref_start_idx, ch.ref_end_idx

            if op == "insert":
                continue

            # for equal/delete/substitute: total increases for ref tokens
            for tok in ref[rs:re]:
                # tok must be a string token
                if isinstance(tok, list):
                    # unexpected nesting; flatten
                    for t in tok:
                        stats.setdefault(t, {"total": 0, "correct": 0})
                        stats[t]["total"] += 1
                        if op == "equal":
                            stats[t]["correct"] += 1
                else:
                    stats.setdefault(tok, {"total": 0, "correct": 0})
                    stats[tok]["total"] += 1
                    if op == "equal":
                        stats[tok]["correct"] += 1

    unique_correct = sum(1 for tok, s in stats.items() if s["correct"] >= correct_at_least)

    print(f"Number of unique tokens correctly predicted at least {correct_at_least} times: {unique_correct}")
    print(f"Total unique tokens in GT: {len(stats)}")

    # stats = {tok: s for tok, s in stats.items() if s["correct"] >= correct_at_least}
    if save_path:
        save_json(save_path, stats)

    
    return stats





def word_frequency_training_set(path, save_path = None, type = 'phoneme'):
    data = load_json(path)
    word_count = {}
    for item in data:
        
        transcription = item["phoneme_presentation"] if type == 'phoneme' else item["text"]
        for token in transcription.split(' '):
            if token not in word_count:
                word_count[token] = 0
            word_count[token] += 1
    if save_path:
        save_json(save_path, word_count)
    return word_count



import numpy as np

def build_vectors(test_stats, train_freq, include_zero=False):
    freqs, recalls = [], []

    for w, s in test_stats.items():
        if s["total"] == 0:
            continue

        recall = s["correct"] / s['total']
        freq = train_freq.get(w, 0)
        

        if not include_zero and freq == 0:
            continue

        freqs.append(freq)
        recalls.append(recall)

    return np.array(freqs), np.array(recalls)


from scipy.stats import pearsonr, spearmanr

unique_word_count_test_set = compute_token_stats_with_jiwer("./result/result-tasa-w2i-lsvsc.json", correct_at_least=1, save_path="./result/unique_word_stats_test_set.json")
unique_word_count_training_set = word_frequency_training_set("../dataset/LSVSC_train_word.json", save_path="./result/unique_word_stats_training_set.json", type = 'word')

freqs, recalls = build_vectors(
    unique_word_count_test_set,
    unique_word_count_training_set,
    include_zero=True   # giống paper
)

pearson_r, _ = pearsonr(freqs, recalls)
spearman_rho, _ = spearmanr(freqs, recalls)

print("Words used:", len(freqs))
print("Pearson r:", pearson_r)
print("Spearman rho:", spearman_rho)




Number of unique tokens correctly predicted at least 1 times: 2440
Total unique tokens in GT: 2957
Words used: 2957
Pearson r: 0.23126171412486463
Spearman rho: 0.3899768691222494


Result based on TASA p, c, word result

## Phoneme
Number of unique tokens correctly predicted at least 1 times: 2474
Total unique tokens in GT: 3001
Words used: 3001
Pearson r: 0.22260085725351292
Spearman rho: 0.32388628887549537

## Character

Number of unique tokens correctly predicted at least 1 times: 2700
Total unique tokens in GT: 3586
Words used: 3586
Pearson r: 0.24836971446922626
Spearman rho: 0.4811725802044207

## Word

Number of unique tokens correctly predicted at least 1 times: 2440
Total unique tokens in GT: 2957
Words used: 2957
Pearson r: 0.23126171412486463
Spearman rho: 0.3899768691222494